### ISSUE TITLE
New event system

### ISSUE THREAD

**Issue by almarklein** [zoofIO/flexx#60]
This issue describes a new event system to replace `flexx.react`. Discussions welcome.

The event system for Flexx should, like the rest of Flexx, be simple and intuitive to use. The current system borrowed some ideas from reactive programming to achieve this. I was initially enthousiastic about RP, but in retrospect I don't think its the right model.

## The current system

The system in `flexx.react` was a fun experiment, and it certainly has some nice features, but it also has some flaws:

* This "reactive programming" model does not work well in all UI situations, in particular applications geared towards visualization. The reason is that some things *are* events, and sometimes you need more information than just a single value.
* The fact that object "properties" are set/get with a function call is not very Pythonic. Even though I've gotten used to the idea, I still commonly make the mistake of using assignment. This is also a hurdle for newcomers.
* Scroll events don't let themselves be mapped to signals trivially; because scroll events are always relative.
* Unnecessary events to initialize signal values (i.e. the `mouse_down` state to False)

The good features can be summarized as:

* Being able to connect to an event using a decorator at the function that is going to handle the event.
* Being able to easily connect to multiple events.
* Dynamism (e.g. connect to `'children.*.size'`).
* An anologue and API compatible implementation in JS.

## The new system

I am planning/proposing a new event system that is somewhat more conventional, in that it uses events, but has the nice features of the current system.

What this new system will have (that the current system has not):

* A better name, as react leans towards RP and is confusing with reactjs.
* Events are handled in the next event loop iteration (unless there is no event loop, in which case events are probably handled synchronously).
* Event handlers can also be called manually (i.e. an event handler remains a callable).

### Generating events

Events can only apply to `HasEvents` objects (you cannot create events on their own, as you could with signals). ~~You can emit an event using `x.emit(name, **values)`. This causes an event object to be created, which is basically a dictionary from `values` with some extra info, like `owner`.~~

Edit: I think it makes sense to allow emitting events doing `x.event_name(ob)`, which will use ob to produce and emit an event object.

### Event objects

Event objects are simply dicts with some fields. For property-change-events, the fields will be 'value', 'previous_value', 'time', 'previous_time', 'owner'. I think it would be useful to create a subclass to allow easier accessing using `ev.value`, but otherwise they're just dicts.

### Connecting to events

You can connect to an event using:
``` python
@connect('foo')
def on_foo(*events):
    ...

@connect('foo', 'bar')
def update(*events):
    ...
```
The decorator is basically the same as it is now, but the function signature now accepts a series of events objects. Even for `on_foo()`, the function can be called with multiple 'foo' events, if multiple events where generated during one event loop iteration. Event handlers are free to process all events or only the last one (as `update()` will likely do). The decorator turns the function into a callable object (maybe even the original function?), so that you can manually call `update()`, which causes `events` to be empty.

### Properties

A decorator `@prop` replaces the `@input` and can be used to create properties. When a property is changed, an event with the same name is emitted, which contains information about the new value, old value, and timings.

Similarly, there will be a `@readprop` (or different name?) decorator for readonly properties (much like the current `@source`.

```python

class Foo(HasEvents):
    @prop
    def name(s='John'):
        return str(s).capitalize()

foo = Foo()

@connect('foo.name')
def on_name_change(*events):
    print('name %r changed to %r' % (events[-1].value, events[0].previous_value)

foo.name = 'X'
foo.name = 'jane'

# in the next event loop, we will see one message saying "name 'John' changed to 'Jane'".
```

### Defining events

Event can be defined using 

```python
class Foo(HasEvents):
    
    # Short
    my_event = Event('x', 'y',  # x and y are names that the event object will have
                                  docs='docs on this event')
   
   # I also like this, its 2 lines more, but it allows more flexibility, it allows
   # emitting events using `foo.my_event(ob)`, and I prefer writing docstrings this way.  
   @event
    def my_event(ob):
        """ docs on this event """
        return Event(x=ob.x, y=ob.y)   # create Event object based on input passed to `emit()`.
```

### Streams and caching

One central concept of RP is that you have streams of signals, where new signals are created from a combination of signals etc. A model like this is still possible, by emitting new events from inside an event handler. We can consider using `yield` to generate new events.

Another concept of RP is caching of signal values, which can help make apps that have a long-running operation to run efficiently. I think that it will take little extra effor to create a proxy event to implement this in the new system.

### Event handling order and propagation

Because multiple events can be handled by a single function, handlers are not called in the order in which they were connected. The order should be deterministic, and I think it makes sense to use the event handler name for this. Event order can be influenced by chosing the handler name, and by a keyword argument to the decorator, e.g. `@connect('foo', 'bar', key='__need this first')`. A handler can not prevent the event being passed to other handlers.

The system itself does not implement any form of event propagation, though systems using it *can* do so. Here's an (overly simplified) example:

```python
class RootNode(Node):
     @connect('somewhere.mouse_down')
     def _on_mouse_event(self, ev): 
         ev.handled = False
         for node in reversed(children):
               node.mouse_event(ev)
               if ev.handled:
                    break

class NodeThatReactsToLMB(Node):
     def mouse_event(self, ev): 
         if ev.button == 1:
             ...
             ev.handled = True

class NodeThatReactsToRMB(Node):
     def mouse_event(self, ev): 
         if ev.button == 2:
             ...
             ev.handled = True


```

### Loose ends

Suggestions and discussion welcome!

* A name for this subpackage. Perhaps something derived of "zanshin"? Or something general enough to represent the ideas of property and event.
* ~~Does each event need a placeholder, or can events exist by virtue of just emitting them?~~
* Name for readonly property.

**Comment by Korijn** [comment]
* Is `HasEvents` as a base class/mixin really a good idea? It sounds like something you'd implement as an interface in static languages; this is usually done with duck typing in Python, right?
* Are string based connectors the best way to go? It does not play well with refactoring. Perhaps _also_ allow the passing of functions?
* You might want to read about jQuery's event system for some inspiration, which supports _namespacing_: https://css-tricks.com/namespaced-events-jquery/

**Comment by almarklein** [comment]
We need a metaclass to allow these kinds of decorators. Or maybe we can do without if we make some restrictions, but we'd need the `__init__` to be run to "prepare" the properties, events and handlers. What kind of use-case were you thinking of to go without a base class? Perhaps something like this could work:
```python
class MyObject:
    def __init__(self):
        events.prepare_object(self)
    @event.prop
    def foo(self, v):
        return float(v)
    
    @event.connect('foo')
    def _handle_foo(self, *events):
         ...
```

I'll read into the jQuery document. Thanks.

**Comment by almarklein** [comment]
About connecting events. In most event systems you bind a handler to the object that generates the event, which would often be in an init. I went with an approach where you specify *at the handler* to what event(s) it is bound. I've found this to lead to clearer code. This does lead to a string-based description of the event, but I think that's ok. An advantage is that it allows binding to events that do not yet exist. And it also allows dynamism (e.g. 'children.*.foo' or 'parent.size'), which I found to be a very powerful mechanism. The downside is that it is more sensitive to typos. For refactoring there would be an equal amount of names to change, though with using a reference you could be helped by static analysis tools. Overall, I think the advantages outweigh the disadvantages.

It would certainly be possible to allow binding by reference too. Preferably I'd stick to Just One (TM) approach though, and add another later when/if its needed. However, one case that pops to mind is the binding to a function (not a method):
```python
# As it is now; this assumes there is an object `h` in the current scope
@event.connect('h.foo')
def handle_foo(*events):
    ...
# Maybe this could be more explicit, something like:
@h.connect('foo')
def handle_foo(*events):
    ...
```
In other words, the string description of an event would always be relative to the a certain HasEvents object.

**Comment by Korijn** [comment]
I see your point. Specifying at the handler is definitely a pleasant way to connect, I agree on that.

The dynamism part is new to me, though, are there reserved words? You could protect against typos by issuing warnings when the specified elements doesn't match anything. Obviously, that shouldn't be done for wildcard selectors though.

**Comment by almarklein** [comment]
The dynamism part basically allows to traverse objects. But also objects that are properties, and get updates when that property changes. E.g. for 'parent.size' you get an event when the size of the current parent changes, and when the parent changes, the event is automatically reconnected, and you get an event. Binding to 'children.*.foo' gets you an event when the 'foo' event on any of the children occurs. Other than perhaps the '*', there are no reserved words.

Yes, I'm leaning to making it more strict and force the connection of the handler to be successful, unless its a dynamism one (e.g. `parent` can be `None`). I think I also like the `@h.connect('foo')` over `event.connect('h.foo')`, because it feels less magical, and maybe we can get rid of all the `sys._getframe()` then.

**Comment by pembeci** [comment]
May be you can even allow the user to specify how much to collect the event values. This way if only the last value is ever used, only that value can be kept preserving system resources. Something like:

```python
@connect('mouse_move', collect=-1)  # collect only the last value
def on_move(event):  # just a single event

@connect('mouse_move', collect=-3)  # collect the last 3 values
def on_move(*events):  

@connect('mouse_move', collect=2)  # collect first 2 values, ignore rest
def on_move(fst, snd):  

# you may even go further/crazier
@connect('mouse_move', collect=(1,-2))  # collect first value, and the last two values
def on_move(fst,prev,last):  
```

**Comment by almarklein** [comment]
@pembeci I doubt that it matters (event objects are pretty lightweight) but this is certainly an idea to keep in mind. Thanks.

**Comment by datnamer** [comment]
http://elm-lang.org/blog/farewell-to-frp

**Comment by pembeci** [comment]
@almarklein Sorry I couldn't respond sooner. I guess I can't still forget the days of 64K RAMs. Whenever I wrote some code where memory usage can be an issue my first instinct is try to optimize things. But you are right, with today's hardware simplicity should have higher priority.

**Comment by Korijn** [comment]
Everyone's moving to subscriptions... flexx, django, elm... :)
